### DISCUSSION TITLE
Scene.enter async function stuck

### DISCUSSION THREAD

**Discussion by ihda06** [telegraf/telegraf#1886] [Q&A] [answered]
Hello, i had problem when i tried to fetch data when user enter one scene, i use scene.enter() function and the bot start stuck when enter the scene its just like waiting for the response and if user triger new event like message or action the function work again but it make double response. when i put async function before enter the scene like using bot.command or bot.on then enter the scene, the function work properly, 

```javascript
welcome.action("TopicGenerator", async (ctx) => {
  await TopicGenerator(ctx) // i put fetch function before enter because if i enter first the bot will be stuck
  ctx.scene.enter("ResultTopik");
});
```
i want bot enter scene first and then fetching data

one more detail, on my local that async function on scene.enter() work properly but when i deploy on cyclic.sh the async function became stuck

anyone have solution for this?

**Comment by MKRhere** [comment]
Post the actual code that had an issue.

**Reply by ihda06** [reply]
Thank you for your response

This is my welcome scene

```javascript
const welcome = new Scene("welcome");

welcome.enter((ctx) => {
  ctx.reply(greeting(ctx.session.state.userInfo.first_name), {
    reply_markup: {
      inline_keyboard: [
        /* One button */
        [{ text: "Topic Generator", callback_data: "topicgenerator" }],
      ],
    },
  });
});

welcome.action("topicgenerator", (ctx) => {
 ctx.scene.enter("topicgenerator")
});

module.exports = {
  welcome,
};
``` 

This is my topicgenerator scene 
```javascript
const topicgenerator = new Scene("topicgenerator");

topicgenerator.enter(async (ctx) => {
  try {
    let rawdata = await Airtables("TopicGenerator").select().all(); //collect data from airtable (Async func) and stuck here
    rawdata = transformData(rawdata);
    const topic = randomizer(rawdata);
    ctx.reply(`This is your topic\n \n💭\n \n${topic.Topic}`, {
      reply_markup: {
        inline_keyboard: [
          /* One button */
          [{ text: "Add topic", callback_data: "addTopic" }],
        ],
      },
    });
  } catch (error) {
    console.log(error);
    ctx.reply("error");
  }
});
``` 

i just wanted fetch data in topicgenerator.enter()

if its still confusing let me know

**Reply by MKRhere** [reply, marked as answer]
EDIT: This is no longer relevant. See replies.

Your code seems okay. I'm assuming you're in webhook mode. I see that `scene.enter` callback becomes immediately unhandled in Telegraf code. cyclic.sh may be terminating the process when this happens. If that's what happens, there's no way to fix this. Your initial instinct to do:

```TS
await TopicGenerator(ctx);
ctx.scene.enter("ResultTopik");
```

Was right. In fact, it runs in this same order when you move it to a `scene.enter` callback, with the exception being the async function is not awaited for anywhere. If you want, you could make a function to be able to enter from multiple places without repeating code:

```TS
async function enterResultTopik(ctx) {
	await TopicGenerator(ctx); // or just the code from scene.enter here
	ctx.scene.enter("ResultTopik");
}
```

And then call it like this:

```TS
welcome.action("TopicGenerator", async (ctx) => {
  await enterResultTopik(ctx);
});
```

**Reply by ihda06** [reply]
Thank u for ur answer so its better use async function like fetching data before entering the scene right?

**Reply by MKRhere** [reply]
Correct.

**Reply by MKRhere** [reply]
Hello, me again. Seems like I was mistaken about an API, and the solution to your original problem is simply to `await ctx.scene.enter()`.

```TS
welcome.action("topicgenerator", async ctx => {
	await ctx.scene.enter("topicgenerator");
});
```

You could also return it for the same effect. Sorry for the confusion!
